arangod/GeneralServer/H2CommTask.cpp

```diff
@@ -0,0 +1,654 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2020 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Simon Grätzer
+////////////////////////////////////////////////////////////////////////////////
+
+#include "H2CommTask.h"
+
+#include "Basics/Exceptions.h"
+#include "Basics/ScopeGuard.h"
+#include "Basics/asio_ns.h"
+#include "Cluster/ServerState.h"
+#include "GeneralServer/GeneralServer.h"
+#include "GeneralServer/GeneralServerFeature.h"
+#include "Logger/LogMacros.h"
+#include "Rest/HttpRequest.h"
+#include "Rest/HttpResponse.h"
+#include "Statistics/ConnectionStatistics.h"
+#include "Statistics/RequestStatistics.h"
+
+#include <velocypack/velocypack-aliases.h>
+#include <cstring>
+
+using namespace arangodb;
+using namespace arangodb::basics;
+using namespace arangodb::rest;
+
+using arangodb::velocypack::StringRef;
+
+template <SocketType T>
+/*static*/ int H2CommTask<T>::on_begin_headers(nghttp2_session* session,
+                                               const nghttp2_frame* frame, void* user_data) {
+  H2CommTask<T>* me = static_cast<H2CommTask<T>*>(user_data);
+
+  if (frame->hd.type != NGHTTP2_HEADERS || frame->headers.cat != NGHTTP2_HCAT_REQUEST) {
+    return 0;
```

Comment:
should we return an error from here?

Reply:
I don't know I am following the example code